nomad/structs/node_pool.go

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package structs
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/hashicorp/go-multierror"
+)
+
+const (
+	// NodePoolAll is a reserved node pool that always includes all nodes in
+	// the cluster.
+	NodePoolAll            = "all"
+	NodePoolAllDescription = "Node pool with all nodes in the cluster."
+
+	// NodePoolDefault is a reserved node pool for nodes that don't specify a
+	// node pool in their configuration.
+	NodePoolDefault            = "default"
+	NodePoolDefaultDescription = "Default node pool."
+
+	// maxNodePoolDescriptionLength is the maximum length allowed for a node
+	// pool description.
+	maxNodePoolDescriptionLength = 256
+)
+
+var (
+	// validNodePoolName is the rule used to validate a node pool name.
+	validNodePoolName = regexp.MustCompile("^[a-zA-Z0-9]{1,128}$")
+)
+
+// NodePool allows partioning infrastructure
+type NodePool struct {
+	// Name is the node pool name. It must be unique.
+	Name string
+
+	// Description is the human-friendly description of the node pool.
+	Description string
+
+	// Meta is a set of user-provided metadata for the node pool.
+	Meta map[string]string
+
+	// SchedulerConfiguration is the scheduler configuration specific to the
+	// node pool.
+	SchedulerConfiguration *NodePoolSchedulerConfiguration
+
+	// Raft indexes.
+	CreateIndex uint64
+	ModifyIndex uint64
+}
+
+// Validate returns an error if the node pool is invalid.
+func (n *NodePool) Validate() error {
+	var mErr *multierror.Error
+
+	if !validNodePoolName.MatchString(n.Name) {
+		mErr = multierror.Append(mErr, fmt.Errorf("invalid name %q, must match regex %s", n.Name, validNodePoolName))
+	}
+	if len(n.Description) > maxNodePoolDescriptionLength {
+		mErr = multierror.Append(mErr, fmt.Errorf("description longer than %d", maxNodePoolDescriptionLength))
+	}
+
+	mErr = multierror.Append(mErr, n.SchedulerConfiguration.Validate())
+
+	return mErr.ErrorOrNil()
+}
+
+// Copy returns a deep copy of the node pool.
+func (n *NodePool) Copy() *NodePool {
+	if n == nil {
+		return nil
+	}
+
+	nc := new(NodePool)
+	*nc = *n
+	nc.SchedulerConfiguration = n.SchedulerConfiguration.Copy()
+
+	return nc
+}
+
+// IsReserved returns true if the node pool is one of the reserved pools.
+// These pools cannot be deleted or modified.
+func (n *NodePool) IsReserved() bool {
+	switch n.Name {
+	case NodePoolAll, NodePoolDefault:
+		return true
+	default:
+		return false
+	}
+}
```

Comment:
Hum...good point. Reserved was used in one of the internal RFC comments, so I just ran with it 😅 

Other names I thought:

* `protected`: maybe not great either because it may suggest it was user protected.
* `system`: as in it's controlled by the system
* `internal`: not good because it sounds like something users shouldn't be using
* `built-in`: probably the least worse option?

Reply:
I went with built-in for now and also removed the user-facing references to this expression.